include/deal.II/base/mpi.h

```diff
@@ -1996,6 +2232,144 @@ namespace Utilities
     }
 
 
+    template <typename T>
+    Future<void>
+    isend(const T &          object,
+          MPI_Comm           communicator,
+          const unsigned int target_rank,
+          const unsigned int mpi_tag)
+    {
+#  ifndef DEAL_II_WITH_MPI
+      Assert(false,
+             ExcMessage(
+               "This function is not useful when called without MPI."));
+      (void)object;
+      (void)communicator;
+      (void)target_rank;
+      (void)mpi_tag;
+      return Future<void>([]() {}, []() {});
+#  else
+      // Create a pointer to a send buffer into which we pack the object
+      // to be sent. The buffer will be released by the Future object once
+      // the send has been verified to have succeeded.
+      //
+      // Conceptually, we would like this send buffer to be a
+      // std::unique_ptr object whose ownership is later handed over
+      // to the cleanup function. That has the disadvantage that the
+      // cleanup object is a non-copyable lambda capture, leading to
+      // awkward semantics. Instead, we use a std::shared_ptr; we move
+      // this shared pointer into the cleanup function, which means
+      // that there is exactly one shared pointer who owns the buffer
+      // at any given time, though the latter is not an important
+      // optimization.
+      std::shared_ptr<std::vector<char>> send_buffer =
+        std::make_unique<std::vector<char>>(Utilities::pack(object, false));
+
+      // Now start the send, and store the result in a request object that
+      // we can then wait for later:
+      MPI_Request request;
+      const int   ierr =
+        MPI_Isend(send_buffer->data(),
+                  send_buffer->size(),
+                  mpi_type_id_for_type<decltype(*send_buffer->data())>,
```

Comment:
This is always a `char`, right?
```suggestion
                  MPI_CHAR)>,
```

Reply:
Yes, but it is brittle to hard code it like this. A better approach is to have the MPI type deduced automatically so that it matches and references the object to be sent, like I do with `mpi_type_id_for_type`. I would like to leave it as is.